journal/src/test/java/io/zeebe/journal/file/SegmentedJournalTest.java

```diff
@@ -0,0 +1,376 @@
+        new Namespace.Builder()
+            .register(Namespaces.BASIC)
+            .nextId(Namespaces.BEGIN_USER_CUSTOM_ID)
+            .register(PersistedJournalRecord.class)
+            .register(UnsafeBuffer.class)
+            .name("Journal")
+            .build();
+
+    entry = "TestData".getBytes();
+    data.wrap(entry);
+
+    final var entrySize =
+        namespace.serialize(new PersistedJournalRecord(1, 1, Integer.MAX_VALUE, data)).length
+            + Integer.BYTES;
+    entriesPerSegment = 10;
+
+    journal =
+        SegmentedJournal.builder()
+            .withDirectory(directory.resolve("data").toFile())
+            .withMaxSegmentSize(entriesPerSegment * entrySize + JournalSegmentDescriptor.BYTES)
+            .withJournalIndexDensity(5)
+            .build();
+  }
+
+  @Test
+  public void shouldBeEmpty() {
+    // when-then
+    assertThat(journal.isEmpty()).isTrue();
+  }
+
+  @Test
+  public void shouldNotBeEmpty() {
+    // given
+    journal.append(1, data);
+
+    // when-then
+    assertThat(journal.isEmpty()).isFalse();
+  }
+
+  @Test
+  public void shouldAppendData() {
+    // given
+    final var recordAppended = journal.append(1, data);
+    assertThat(recordAppended.index()).isEqualTo(1);
+
+    // when
+    final var recordRead = journal.openReader().next();
+
+    // then
+    assertThat(recordAppended.index()).isEqualTo(recordRead.index());
+    assertThat(recordAppended.asqn()).isEqualTo(recordRead.asqn());
+    assertThat(recordAppended.checksum()).isEqualTo(recordRead.checksum());
+  }
+
+  @Test
+  public void shouldAppendMultipleRecords() {
+    for (int i = 0; i < 10; i++) {
+      final var recordAppended = journal.append(i + 10, data);
+      assertThat(recordAppended.index()).isEqualTo(i + 1);
+    }
+
+    final var reader = journal.openReader();
+    for (int i = 0; i < 10; i++) {
+      assertThat(reader.hasNext()).isTrue();
+      final var recordRead = reader.next();
+      assertThat(recordRead.index()).isEqualTo(i + 1);
+      final byte[] data = new byte[recordRead.data().capacity()];
+      recordRead.data().getBytes(0, data);
+      assertThat(recordRead.asqn()).isEqualTo(i + 10);
+      assertThat(data).containsExactly(entry);
+    }
+  }
+
+  @Test
+  public void shouldAppendAndReadMultipleRecords() {
+    final var reader = journal.openReader();
+    for (int i = 0; i < 10; i++) {
+      // given
+      entry = ("TestData" + i).getBytes();
+      data.wrap(entry);
+
+      // when
+      final var recordAppended = journal.append(i + 10, data);
+      assertThat(recordAppended.index()).isEqualTo(i + 1);
+
+      // then
+      assertThat(reader.hasNext()).isTrue();
+      final var recordRead = reader.next();
+      assertThat(recordRead.index()).isEqualTo(i + 1);
+      final byte[] data = new byte[recordRead.data().capacity()];
+      recordRead.data().getBytes(0, data);
+      assertThat(recordRead.asqn()).isEqualTo(i + 10);
+      assertThat(data).containsExactly(entry);
+    }
```

Comment:
I would splitt this. First append as when and then I can read all

Reply:
I will split this. But shouldn't we also test this scenario where reading while writing?

Reply:
yeah makes probably sense